File: backend/app/schemas/replay.py
"""
策略回放 Pydantic Schema
PRD 4.17 策略回放功能
"""

from datetime import date, datetime
from decimal import Decimal
from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field


class ReplaySpeed(str, Enum):
    """回放速度"""
    HALF = "0.5x"
    NORMAL = "1x"
    DOUBLE = "2x"
    FAST = "5x"


class ReplayStatus(str, Enum):
    """回放状态"""
    IDLE = "idle"
    PLAYING = "playing"
    PAUSED = "paused"


# ============ 回放配置 ============


class ReplayConfig(BaseModel):
    """回放配置"""
    strategy_id: str = Field(..., description="策略ID")
    symbol: str = Field(..., description="股票代码")
    start_date: date = Field(..., description="开始日期")
    end_date: date = Field(..., description="结束日期")
    speed: ReplaySpeed = Field(default=ReplaySpeed.NORMAL, description="回放速度")


class ReplayState(BaseModel):
    """回放状态"""
    session_id: str = Field(..., description="会话ID")
    config: ReplayConfig
    status: ReplayStatus = ReplayStatus.IDLE
    current_time: datetime
    current_bar_index: int = 0
    total_bars: int = 0

    # 模拟持仓
    position_quantity: int = 0
    position_avg_cost: Decimal = Decimal("0")
    cash: Decimal = Decimal("100000")

    # 回放统计
    total_signals: int = 0
    executed_signals: int = 0
    total_return_pct: float = 0
    benchmark_return_pct: float = 0


# ============ 历史数据 ============


class HistoricalBar(BaseModel):
    """历史K线"""
    timestamp: datetime
    open: Decimal
    high: Decimal
    low: Decimal
    close: Decimal
    volume: int


class ThresholdConfig(BaseModel):
    """阈值配置"""
    value: float
    direction: str  # above/below
    passed: bool


class FactorSnapshot(BaseModel):
    """因子快照"""
    timestamp: datetime
    factor_values: dict[str, float] = Field(default_factory=dict)
    thresholds: dict[str, ThresholdConfig] = Field(default_factory=dict)
    overall_signal: str = "hold"  # buy/sell/hold
    conditions_met: int = 0
    conditions_total: int = 0


class SignalEvent(BaseModel):
    """信号事件"""
    event_id: str
    timestamp: datetime
    event_type: str  # buy_trigger, sell_trigger, condition_check
    symbol: str
    price: Decimal
    description: str
    factor_details: Optional[dict] = None


# ============ 回放响应 ============


class SignalMarker(BaseModel):
    """信号标记"""
    index: int
    time: str
    type: str  # buy/sell


class ReplayInitResponse(BaseModel):
    """回放初始化响应"""
    state: ReplayState
    total_bars: int
    signal_markers: list[SignalMarker] = Field(default_factory=list)


class ReplayTickResponse(BaseModel):
    """回放Tick响应"""
    state: ReplayState
    bar: HistoricalBar
    factor_snapshot: FactorSnapshot
    events: list[SignalEvent] = Field(default_factory=list)


class ReplayInsight(BaseModel):
    """回放洞察"""
    total_signals: int
    execution_rate: float
    win_rate: float
    alpha: float

    # AI洞察
    ai_insights: list[str] = Field(default_factory=list)

    # 收益对比
    strategy_return: float
    benchmark_return: float


class ReplayExport(BaseModel):
    """回放导出"""
    events: list[SignalEvent]
    summary: ReplayInsight


# ============ 请求模型 ============


class SeekRequest(BaseModel):
    """跳转请求"""
    target_time: datetime


class SpeedRequest(BaseModel):
    """速度设置请求"""
    speed: ReplaySpeed

File: frontend/src/types/replay.ts
/**
 * 策略回放类型定义
 * PRD 4.17 策略回放功能
 */

export type ReplaySpeed = '0.5x' | '1x' | '2x' | '5x'
export type ReplayStatus = 'idle' | 'playing' | 'paused'

export interface ReplayConfig {
  strategyId: string
  symbol: string
  startDate: string
  endDate: string
  speed: ReplaySpeed
}

export interface ReplayState {
  sessionId: string
  config: ReplayConfig
  status: ReplayStatus
  currentTime: string
  currentBarIndex: number
  totalBars: number

  // 模拟持仓
  positionQuantity: number
  positionAvgCost: number
  cash: number

  // 统计
  totalSignals: number
  executedSignals: number
  totalReturnPct: number
  benchmarkReturnPct: number
}

export interface HistoricalBar {
  timestamp: string
  open: number
  high: number
  low: number
  close: number
  volume: number
}

export interface ThresholdConfig {
  value: number
  direction: 'above' | 'below'
  passed: boolean
}

export interface FactorSnapshot {
  timestamp: string
  factorValues: Record<string, number>
  thresholds: Record<string, ThresholdConfig>
  overallSignal: 'buy' | 'sell' | 'hold'
  conditionsMet: number
  conditionsTotal: number
}

export interface SignalEvent {
  eventId: string
  timestamp: string
  eventType: 'buy_trigger' | 'sell_trigger' | 'condition_check'
  symbol: string
  price: number
  description: string
  factorDetails?: Record<string, any>
}

export interface SignalMarker {
  index: number
  time: string
  type: 'buy' | 'sell'
}

export interface ReplayInitResponse {
  state: ReplayState
  totalBars: number
  signalMarkers: SignalMarker[]
}

export interface ReplayTickResponse {
  state: ReplayState
  bar: HistoricalBar
  factorSnapshot: FactorSnapshot
  events: SignalEvent[]
}

export interface ReplayInsight {
  totalSignals: number
  executionRate: number
  winRate: number
  alpha: number
  aiInsights: string[]
  strategyReturn: number
  benchmarkReturn: number
}

// 颜色配置
export const REPLAY_COLORS = {
  // K线区域
  replayedBars: 'normal',
  currentBar: '#8b5cf6',
  futureBars: 'rgba(128,128,128,0.3)',
  positionLine: '#8b5cf6',

  // 信号标记
  buySignal: '#22c55e',
  sellSignal: '#ef4444',

  // 因子状态
  factorPassed: '#22c55e',
  factorFailed: '#ef4444',
}

// 速度配置
export const REPLAY_SPEEDS: { value: ReplaySpeed; label: string; interval: number }[] = [
  { value: '0.5x', label: '0.5x 慢速', interval: 2000 },
  { value: '1x', label: '1x 正常', interval: 1000 },
  { value: '2x', label: '2x 快速', interval: 500 },
  { value: '5x', label: '5x 极速', interval: 200 },
]

// 格式化时间
export function formatReplayTime(timestamp: string): string {
  const date = new Date(timestamp)
  return date.toLocaleString('zh-CN', {
    month: '2-digit',
    day: '2-digit',
    hour: '2-digit',
    minute: '2-digit',
  })
}

// 格式化百分比
export function formatPercent(value: number): string {
  const sign = value >= 0 ? '+' : ''
  return `${sign}${(value * 100).toFixed(2)}%`
}

// 获取信号图标
export function getSignalIcon(type: string): string {
  switch (type) {
    case 'buy':
    case 'buy_trigger':
      return '🟢'
    case 'sell':
    case 'sell_trigger':
      return '🔴'
    default:
      return '⚪'
  }
}

// 获取信号标签
export function getEventLabel(type: string): string {
  switch (type) {
    case 'buy_trigger':
      return '买入信号触发'
    case 'sell_trigger':
      return '卖出信号'
    case 'condition_check':
      return '条件检查'
    default:
      return '未知事件'
  }
}
